### PR TITLE
Feature: Save mirror

### PR DESCRIPTION
**Pull request by maikthomas**
So far I've added the front-end logic for the client-mirror - I figured let's get this one right first then replicate for the server.

I've added the back-end api paths and dummy responses. Need to add integration with Redis and actual save/retrieve functionality.

### CODE REVIEW

File: src/js/App.js [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
-      <Route exact path="/" component={ClientMirror} />
-      <Route path="/server-mirror" component={ServerMirror} />
-    </Layout>
-  </HashRouter>,
+      <Layout>
+        <Switch>
+          <Redirect exact from="/" to="client-mirror" />
+          <Route path="/client-mirror/:mirrorId?" component={ClientMirror} />
+          <Route path="/server-mirror/:mirrorId?" component={ServerMirror} />
+        </Switch>
```
Comment (maikthomas) [inline src/js/App.js:15]: If you don't use `<Switch>` then you can't use `exact` in the redirect, which leads to things things `/client-mirror/` redirecting to `/client-mirror/client-mirror` and so and and so forth for eternity :|
https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/Switch.md

File: src/js/components/clientMirror/ActionBar.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import React from 'react';
 import PropTypes from 'prop-types';
+import FaFloppyO from 'react-icons/lib/fa/floppy-o';
```
Comment (maikthomas) [inline src/js/components/clientMirror/ActionBar.js:3]: Just a quick solution, perhaps we want to use something else?

File: src/js/components/clientMirror/ActionBar.js [hunk 3]
```diff
@@ -60,6 +61,13 @@ class ActionBar extends React.Component {
             className="glyphicon glyphicon-refresh spinning"
           />
         </button>
+        <button className="btn" onClick={() => {this.props.onSaveClick()}}>
+          <FaFloppyO /> Save
+          <span
+            style={{ display: this.props.options.isRunning ? 'inline-block' : 'none' }}
```
Comment (maikthomas) [inline src/js/components/clientMirror/ActionBar.js:67]: I think we need to sort out the style. This works fine on a big screen but on my laptop screen it goes over onto two lines

File: src/js/pages/ClientMirror.js [hunk 4]
```diff
@@ -108,4 +156,5 @@ class ClientMirror extends React.Component {
   }
 }
 
-export default ClientMirror;
+
+export default withRouter(ClientMirror);
```
Comment (maikthomas) [inline src/js/pages/ClientMirror.js:160]: these `withRouter` part are needed to do the redirect. 
https://reacttraining.com/react-router/web/api/withRouter

File: src/js/App.js (hunk 1 above)
Comment (marcioaffonso) [reply src/js/App.js:15]: Great!

File: src/js/components/clientMirror/ActionBar.js (hunk 2 above)
Comment (marcioaffonso) [reply src/js/components/clientMirror/ActionBar.js:3]: Yes, I think we should Bootstrap icons as we are already using them: https://www.w3schools.com/icons/bootstrap_icons_glyphicons.asp

I can change that.

File: src/js/pages/ClientMirror.js (hunk 4 above)
Comment (marcioaffonso) [reply src/js/pages/ClientMirror.js:160]: Great!

File: src/js/App.js [hunk 5]
```diff
@@ -1,16 +1,22 @@
...
+      </Layout>
+    </HashRouter>,
   document.getElementById('appContainer')
 );
+
+const NotFound = () => (
```
Comment (marcioaffonso) [inline src/js/App.js]: Does this work just by assigning this constant?

File: src/js/components/clientMirror/ActionBar.js (hunk 3 above)
Comment (marcioaffonso) [reply src/js/components/clientMirror/ActionBar.js:67]: I can change that too, don't worry.

File: src/js/pages/ClientMirror.js [hunk 6]
```diff
@@ -1,21 +1,50 @@
...
 
 const initialState = require('../data/client-mirror-initial-state');
 const jsSdkVersions = require('../data/js-sdk-versions');
 
 class ClientMirror extends React.Component {
-  constructor() {
-    super();
+  constructor(props) {
+    super(props);
     this.state = {
```
Comment (marcioaffonso) [inline src/js/pages/ClientMirror.js:14]: I think this code is duplicated in `componentWillReceiveProps`. Maybe the constructor could be removed.

**Review by marcioaffonso — CHANGES_REQUESTED** [review]

File: src/js/pages/ClientMirror.js (hunk 6 above)
Comment (maikthomas) [reply src/js/pages/ClientMirror.js:14]: I'll investigate this.

File: src/js/components/clientMirror/ActionBar.js (hunk 2 above)
Comment (maikthomas) [reply src/js/components/clientMirror/ActionBar.js:3]: Great, do you want to just push it onto this branch or change it after?

File: src/js/App.js (hunk 5 above)
Comment (maikthomas) [reply src/js/App.js]: No. This is a mistake

File: src/js/pages/ClientMirror.js (hunk 6 above)
Comment (maikthomas) [reply src/js/pages/ClientMirror.js:14]: @marcioaffonso Here it says `constructor` comes under Mounting: "Mounting
These methods are called when an instance of a component is being created and inserted into the DOM"
Whilst `componentWIllReceiveProps` comes under Updating: "An update can be caused by changes to props or state. These methods are called when a component is being re-rendered:"
and below "React doesn't call componentWillReceiveProps with initial props during mounting"

So I think we need both. Source:
https://facebook.github.io/react/docs/react-component.html#the-component-lifecycle

**Review by marcioaffonso — APPROVED** [review]

**Pull request merged**